floss/language/go/extract.py

```diff
@@ -0,0 +1,318 @@
+from floss.results import StaticString, StringEncoding
+
+logger = logging.getLogger(__name__)
+
+MIN_STR_LEN = 6
+
+
+def extract_stackstrings(extract_stackstring_pattern, section_data, min_length) -> List[StaticString]:
+    stack_strings = list()
+    for m in extract_stackstring_pattern.finditer(section_data):
+        for i in range(1, 8):
+            try:
+                binary_string = m.group(i)
+                if not binary_string:
+                    continue
+
+                addr = m.start()
+                try:
+                    string = StaticString.from_utf8(binary_string, addr, min_length)
+                    stack_strings.append(string)
+                except ValueError:
+                    pass
+            except AttributeError:
+                break
+
+    return stack_strings
+
+
+def xrefs_in_text_segment(
+    pe: pefile.PE, text_segment_data, text_segment_va, start_address, end_address, arch
+) -> List[int]:
+    """
+    Find cross-references to a string in the .text segment.
+    All cross-references are of the form:
+    AMD64:
+    .text:0000000000408389 48 8D 05 80 08 0C 00            lea     rax, unk_4C8C10
+    .text:00000000004736F0 4C 8D 05 84 47 03 00            lea     r8, unk_4A7E7B
+
+    386:
+    .text:004806D2 8D 05 EC 1D 4A 00                       lea     eax, unk_4A1DEC
+
+    """
+    text_segment_xrefs = list()
+
+    if arch == "amd64":
+        text_regex = re.compile(b"(\x48|\x4C)\x8D(?=.(?P<offset>....))", re.DOTALL)
+        for match in text_regex.finditer(text_segment_data):
+            offset = struct.unpack("<I", match.group("offset"))[0]
+            address = text_segment_va + match.start() + offset + 7 + pe.OPTIONAL_HEADER.ImageBase
+            if start_address <= address <= end_address:
+                text_segment_xrefs.append(address)
+    else:
+        text_regex = re.compile(b"\x8D(?=.(?P<offset>....))", re.DOTALL)
+        for match in text_regex.finditer(text_segment_data):
+            offset = struct.unpack("<I", match.group("offset"))[0]
+            address = offset
+            if start_address <= address <= end_address:
+                text_segment_xrefs.append(address)
+
+    return text_segment_xrefs
+
+
+def xrefs_in_rdata_data_segment(section_data, start_address, end_address, arch) -> List[int]:
+    """
+    Find cross-references to a string in the .rdata segment.
+    All cross-references are of the form:
+    00000000004C9D00  19 8C 4A 00 00 00 00 00  0A 00 00 00 00 00 00 00  ..J.............
+    """
+
+    if arch == "amd64":
+        size = 0x10
+        fmt = "<QQ"
+    else:
+        size = 0x8
+        fmt = "<II"
+
+    xrefs_in_rdata_data_segment = list()
+
+    for addr in range(0, len(section_data) - size // 2, size // 2):
+        curr = section_data[addr : addr + size]
+        s_off, s_size = struct.unpack_from(fmt, curr)
+
+        if not (1 <= s_size < 128):
+            continue
+
+        if start_address <= s_off <= end_address:
```

Comment:
i think we should be able to validate that the data at the given address, with the given length, correctly decodes as UTF-8.

i think we should probably also return the length field from this routine, since we can recover it with high confidence. while we may not able to recover it from the code xrefs, we can here, so lets use it to make our algorithm more robust.